python/adbc_driver_manager/adbc_driver_manager/_lib.pyx

```diff
@@ -1481,3 +1483,47 @@ cdef class AdbcStatement(_AdbcHandle):
 cdef const CAdbcError* PyAdbcErrorFromArrayStream(
     CArrowArrayStream* stream, CAdbcStatusCode* status):
     return AdbcErrorFromArrayStream(stream, status)
+
+
+cdef extern from "_blocking_impl.h" nogil:
+    ctypedef void (*BlockingCallback)(void*) noexcept nogil
+    c_string CInitBlockingCallback"pyadbc_driver_manager::InitBlockingCallback"()
+    void CSetBlockingCallback"pyadbc_driver_manager::SetBlockingCallback"(BlockingCallback, void* data)
+    void CClearBlockingCallback"pyadbc_driver_manager::ClearBlockingCallback"()
+
+
+@functools.cache
+def _init_blocking_call():
+    error = bytes(CInitBlockingCallback()).decode("utf-8")
+    if error:
+        raise RuntimeError(error)
+
+
+def _blocking_call(func, args, kwargs, cancel):
+    """
+    Run functions that are expected to block with a native SIGINT handler.
+
+    Parameters
+    ----------
+    """
+    if threading.current_thread() is not threading.main_thread():
+        return func(*args, **kwargs)
+
+    _init_blocking_call()
+
+    # Set the callback for the background thread and save the signal handler
+    CSetBlockingCallback(&_handle_blocking_call, <void*>cancel)
+
+    try:
+        return func(*args, **kwargs)
+    finally:
+        # Restore the signal handler
+        CClearBlockingCallback()
+
+
+cdef void _handle_blocking_call(void* c_cancel) noexcept nogil:
+    # TODO: if this throws, we could save and restore the traceback later
+    # TODO: we could record that this was hit and raise a KeyboardInterrupt above
```

Comment:
Not sure what Python users would prefer - currently it passes through the ADBC exception, not KeyboardInterrupt